File: week07/2/Prob2.py
#User function Template for python3

class Solution:
    #Function to sort a list using quick sort algorithm.
    def quickSort(self,arr,low,high):
        # code here
        if len(arr) == 1:
            return arr
        
        if low < high:
            # pi: pivot value index(partitioning)
            pi = self.partition(arr, low, high)
            self.quickSort(arr, low, pi - 1)
            self.quickSort(arr, pi + 1, high)

        
    
    def partition(self,arr,low,high):
        # code here
        # pivot : last value of arr
        pivot = arr[high]
        i = low

        for j in range(low, high):
            # if current value is smaller than or equal to pivot
            if arr[j] <= pivot:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
        arr[i], arr[high] = arr[high], arr[i]
        return i
        
# Driver code starts, Initial template for Python 3
if __name__ == "__main__":
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().split()))
        Solution().quickSort(arr, 0, n-1)
        for i in range(n):
            print(arr[i], end= " ")
        print()

File: week07/5/prob2.cpp
#include <iostream>
using namespace std;

int partition (int arr[], int low, int high);

//Function to sort an array using quick sort algorithm.
void quickSort(int arr[], int low, int high)
{
    // code here
    // base case : no nothing
    if (low >= high);
    // recursive case
    else {
        int pivot = partition(arr, low, high);
        // cout << low << " " << pivot << " " << high << endl;
        quickSort(arr, low, pivot-1);
        quickSort(arr, pivot+1, high);
    }
}
    
int partition (int arr[], int low, int high)
{
    // Your code here
    int pivot = high;
    int temp;
    for (int curr=low; curr<high; curr++) {
        if (curr < pivot) {
            if (arr[curr] > arr[pivot]) {
                temp = arr[curr];
                arr[curr] = arr[pivot];
                arr[pivot] = temp;
                pivot = curr;
            }
        } else {
            if (arr[curr] < arr[pivot]) {
                temp = arr[curr];
                arr[curr] = arr[pivot+1];
                arr[pivot+1] = arr[pivot];
                arr[pivot] = temp;
                pivot = pivot + 1;
            }
        }
    }
    return pivot;
}
